lib/location/rest/rest_here_wifi.c

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <modem/lte_lc.h>
+#include <cJSON.h>
+#include <cJSON_os.h>
+
+#include <logging/log.h>
+
+#include <cJSON.h>
+
+#include <net/rest_client.h>
+
+#include "rest_here_wifi.h"
+
+LOG_MODULE_DECLARE(location, CONFIG_LOCATION_LOG_LEVEL);
+
+#define HOSTNAME	CONFIG_LOCATION_METHOD_WIFI_SERVICE_HERE_HOSTNAME
+
+BUILD_ASSERT(sizeof(HOSTNAME) > 1, "Hostname must be configured");
+BUILD_ASSERT(
+	sizeof(CONFIG_LOCATION_METHOD_WIFI_SERVICE_HERE_API_KEY) > 1, "API key must be configured");
+#define AUTHENTICATION	"apiKey="CONFIG_LOCATION_METHOD_WIFI_SERVICE_HERE_API_KEY
+
+#define API_LOCATE_PATH		"/v2/locate"
+#define API_KEY_PARAM		"apiKey="CONFIG_LOCATION_METHOD_WIFI_SERVICE_HERE_API_KEY
+#define REQUEST_URL		API_LOCATE_PATH"?"API_KEY_PARAM
+
+#define HEADER_CONTENT_TYPE    "Content-Type: application/json\r\n"
+
+#define HERE_WIFI_POS_JSON_KEY_WLAN "wlan"
+#define HTTPS_PORT 443
+
+/******************************************************************************/
+/* Based on:
+ * https://developer.here.com/documentation/positioning-api/dev_guide/topics/construct-locate-request.html
+ */
+
+static int here_wifi_rest_pos_req_json_format(
+	const struct wifi_scanning_result_info scanning_results[],
+	uint8_t wifi_scanning_result_count,
+	cJSON * const req_obj_out)
+{
+	cJSON *wifi_array = NULL;
+	cJSON *wifi_info_obj = NULL;
+
+	if (!scanning_results || !wifi_scanning_result_count || !req_obj_out) {
+		return -EINVAL;
+	}
+	wifi_array = cJSON_AddArrayToObjectCS(req_obj_out, HERE_WIFI_POS_JSON_KEY_WLAN);
+	if (!wifi_array) {
+		goto cleanup;
+	}
+
+	for (size_t i = 0; i < wifi_scanning_result_count; ++i) {
+		const struct wifi_scanning_result_info current_result = scanning_results[i];
+
+		wifi_info_obj = cJSON_CreateObject();
+
+		if (!cJSON_AddStringToObjectCS(wifi_info_obj, "mac", current_result.mac_addr_str)) {
+			goto cleanup;
+		}
+
+		if (!cJSON_AddNumberToObjectCS(wifi_info_obj, "rss", current_result.rssi)) {
+			goto cleanup;
+		}
+
+		if (!cJSON_AddItemToArray(wifi_array, wifi_info_obj)) {
+			cJSON_Delete(wifi_info_obj);
+			goto cleanup;
+		}
+	}
+	return 0;
+
+cleanup:
+	/* Only need to delete the wifi_array since all items (if any) were added to it */
+	cJSON_DeleteItemFromObject(req_obj_out, HERE_WIFI_POS_JSON_KEY_WLAN);
+	LOG_ERR("Failed to format Here wifi location request, out of memory");
```

Comment:
We should probably use HERE instead of here or Here in wherever we write HERE service name.